client/keys.go

```diff
@@ -450,6 +570,170 @@ func (k *Key) Unseal(in *pb.SealedBytes, opts UnsealOpts) ([]byte, error) {
+		if err = tpm2direct.Unmarshal(in.GetTicket(), &ticket); err != nil {
+			return nil, fmt.Errorf("ticket unmarshal failed: %w", err)
+		}
+
+		creationHash := SessionHashAlg.New()
+		creationHash.Write(in.GetCreationData())
+
+		certifyCreation := tpm2direct.CertifyCreation{
+			SignHandle:   tpm.RHNull,
+			ObjectHandle: sealed,
+			CreationHash: tpm2b.Digest{
+				Buffer: creationHash.Sum(nil),
+			},
+			CreationTicket: ticket,
+		}
+		_, certErr := certifyCreation.Execute(k.transportTPM())
+
+		// There is a bug in some older TPMs, where they are unable to
+		// CertifyCreation when using a Null signing handle (despite this
+		// being allowed by all versions of the TPM spec). To work around
+		// this bug, we use a temporary signing key and ignore the signed
+		// result. To reduce the cost of this workaround, we use a cached
+		// ECC signing key.
+		// We can detect this bug, as it triggers a RCInsufficient
+		// Unmarshaling error.
+		if paramErr, ok := certErr.(tpm2.ParameterError); ok && paramErr.Code == tpm2.RCInsufficient {
+			signer, err := AttestationKeyECC(k.rw)
+			if err != nil {
+				return nil, fmt.Errorf("failed to create fallback signing key: %w", err)
+			}
+			defer signer.Close()
+
+			certifyCreation := tpm2direct.CertifyCreation{
+				SignHandle: tpm2direct.AuthHandle{
+					Handle: tpm.Handle(signer.handle),
+					Auth:   tpm2direct.PasswordAuth(nil),
+				},
+				ObjectHandle: sealed,
+				CreationHash: tpm2b.Digest{
+					Buffer: creationHash.Sum(nil),
+				},
+				CreationTicket: ticket,
+			}
+			_, certErr = certifyCreation.Execute(k.transportTPM())
+
+		}
+		if certErr != nil {
+			return nil, fmt.Errorf("failed to certify creation: %w", certErr)
+		}
+
+		// verify certify PCRs haven't been modified
+		var decodedCreationData tpms.CreationData
+		if err = tpm2direct.Unmarshal(in.GetCreationData(), &decodedCreationData); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal creation data: %w", err)
+		}
+		if !internal.SamePCRSelectionDirect(in.GetCertifiedPcrs(), decodedCreationData.PCRSelect) {
+			return nil, fmt.Errorf("certify PCRs does not match the PCR selection in the creation data")
+		}
+		expectedDigest := internal.PCRDigest(in.GetCertifiedPcrs(), SessionHashAlg)
+		if subtle.ConstantTimeCompare(decodedCreationData.PCRDigest.Buffer, expectedDigest) == 0 {
+			return nil, fmt.Errorf("certify PCRs digest does not match the digest in the creation data")
+		}
+	}
+
+	const sizeOfPCRSelect = 3
+	PCRs := make([]byte, sizeOfPCRSelect)
+	for _, n := range in.GetPcrs() {
+		n := int(n)
+		if n >= 8*sizeOfPCRSelect {
+			return nil, fmt.Errorf("pcr index %d is out of range (exceeds maximum value %d)", n, 8*sizeOfPCRSelect-1)
+		}
+		byteNum := n / 8
+		bytePos := byte(1 << (n % 8))
+		PCRs[byteNum] |= bytePos
+	}
+
+	hash := tpm.AlgID(in.GetHash())
+	sel := tpml.PCRSelection{
+		PCRSelections: []tpms.PCRSelection{
+			{
+				Hash:      hash,
+				PCRSelect: PCRs,
+			},
+		},
+	}
```

Comment:
this should be an unexported util function (e.g., `createTPMLPCRSelection` returning tpml.PCRSelection) in go-tpm-tools/direct since we will likely use it elsewhere and users will definitely want it at some point. Ideally, this is supported in go-tpm in a nice way, but that is out of scope for this PR.

Reply:
I added temporary unexported util functions in keys.go for tpml and tpms PCRSelections.